src/setup_remote.py

```diff
@@ -0,0 +1,101 @@
+    if not os.path.exists(seckey):
+        subprocess.check_call(["gsutil", "cp", "gs://getzlab-secrets/github-service-account/github", seckey])
+
+def get_current_project():
+    ## Assuming on a GCE instance
+    project = subprocess.check_output(
+        "curl -s -H 'Metadata-Flavor: Google' http://metadata.google.internal/computeMetadata/v1/project/project-id",
+        shell=True
+    )
+    project = project.decode()
+    return project
+
+def get_current_instance_name():
+    name = subprocess.check_output(
+        "curl -s -H 'Metadata-Flavor: Google' http://metadata.google.internal/computeMetadata/v1/instance/name",
+        shell=True
+    )
+    name = name.decode()
+    return name
+
+def get_current_zone():
+    ## Assuming on a GCE instance
+    zone = subprocess.check_output(
+        "curl -s -H 'Metadata-Flavor: Google' http://metadata.google.internal/computeMetadata/v1/instance/zone",
+        shell=True
+    )
+    zone = zone.decode()
+    zone = re.sub(".*/", "", zone)
+    return zone
+
+def create_instance(instance_name, project=None, zone=None, machine_type="n1-standard-4", boot_disk_size=200):
+    if project is None:
+        project = get_current_project()
+    if zone is None:
+        zone = get_current_zone()
+    ## I used "enable-oslogin=TRUE", which ensures consistent UID within a project.
+    subprocess.check_call(
+        f"gcloud compute instances create {instance_name} \
+            --quiet \
+            --project {project} \
+            --zone {zone} \
+            --machine-type {machine_type} \
+            --image ubuntu-minimal-2004-focal-v20210511 \
+            --image-project ubuntu-os-cloud \
+            --boot-disk-size {boot_disk_size}GB \
+            --boot-disk-type pd-standard \
+            --metadata=enable-oslogin=TRUE \
+            --scopes cloud-platform,compute-rw"
+            #--metadata-from-file startup-script={startup_script},shutdown-script={shutdown_script} \
+            , shell=True
+    )
+
+    ## Waiting for ssh working
+    n = 0
+    while True:
+        time.sleep(6)
+        try:
+            subprocess.check_call(f"gcloud compute ssh --project {project} --quiet {instance_name} --zone {zone} --command 'true'", shell=True, stderr=subprocess.DEVNULL)
+            break
+        except Exception as err:
+            if n >= 5:
+                raise err
+            pass
+        finally:
+            n += 1
+
+    ## Copy stuff and run user script
+    subprocess.check_call(
+        f"gcloud compute scp {SLURM_GCP_DOCKER_DIR}/ {instance_name}:slurm_gcp_docker --zone {zone} --recurse --project {project} --quiet --scp-flag='-q'", shell=True
+    )
+
+    ## Execute setup script
+    subprocess.check_call(
+        f'gcloud compute ssh {instance_name} --zone {zone} --project {project} -- -o "StrictHostKeyChecking no" -o "UserKnownHostsFile /dev/null" -T "bash ~/slurm_gcp_docker/src/setup_remote_user_startup_script.sh"', shell=True
+    )
```

Comment:
Why not as an instance startup script, rather than manually? Instead of scp'ing the Slurm GCP docker dir from the local machine, why not clone it directly from GitHub in the setup script?

Reply:
The main issue with instance startup script is that we need a way to copy getzlab private ssh key to the machine, and it will be used in the startup script later. If we use instance startup script, we still have to scp the key, detect the key has been copied, and run script that depends on the key. It seems to be easier to just `scp` and `ssh` execute the script in one shot.

Reason to `scp` the entire `slurm_gcp_docker` instead of just the ssh key is that startup script in the remote repo is not guaranteed to be same as the one in local directory when you are making changes (e.g. at this point).